File: app/src/main/java/com/lasercyber/lws/ui/common/upgrade/FirmwareUpgradeCoordinator.java
package com.lasercyber.lws.ui.common.upgrade;

/**
 * Serializes control-card firmware OTA between bundled-home-screen and {@code UpgradeActivity} paths.
 */
public final class FirmwareUpgradeCoordinator {

    private static volatile boolean bundledUpgradeInProgress;
    private static volatile boolean otaUpgradeInProgress;

    private FirmwareUpgradeCoordinator() {
    }

    public static boolean isBundledUpgradeInProgress() {
        return bundledUpgradeInProgress;
    }

    public static boolean isOtaUpgradeInProgress() {
        return otaUpgradeInProgress;
    }

    public static boolean isBusy() {
        return bundledUpgradeInProgress || otaUpgradeInProgress;
    }

    public static boolean canStartFirmwareUpgrade() {
        return !isBusy();
    }

    /**
     * Whether {@code UpgradeActivity} may start Modbus firmware transfer from an OTA ZIP {@code .bin}.
     * The OTA flow sets {@link #otaUpgradeInProgress} for the whole activity; that must not block this call.
     */
    public static boolean canStartOtaFirmwareTransfer() {
        return !bundledUpgradeInProgress;
    }

    public static void markBundledUpgradeStarted() {
        bundledUpgradeInProgress = true;
    }

    public static void markBundledUpgradeEnded() {
        bundledUpgradeInProgress = false;
    }

    public static void setOtaUpgradeInProgress(boolean inProgress) {
        otaUpgradeInProgress = inProgress;
    }
}

File: app/lws_hmi/lib/features/bundled_firmware/application/firmware_upgrade_coordinator.dart
/// Serializes control-board Modbus flash, camera CGI flash, and whole-device OTA.
abstract final class FirmwareUpgradeCoordinator {
  static bool _bundledInProgress = false;
  static bool _otaInProgress = false;

  static bool get isBundledUpgradeInProgress => _bundledInProgress;
  static bool get isOtaUpgradeInProgress => _otaInProgress;
  static bool get isBusy => _bundledInProgress || _otaInProgress;

  static bool canStartFirmwareUpgrade() => !isBusy;

  /// Control-board / camera bundled flash — blocked while whole-device OTA is active.
  static bool canStartBundledFirmwareUpgrade() => canStartFirmwareUpgrade();

  /// Future OTA path: blocked only while bundled is active.
  static bool canStartOtaFirmwareTransfer() => !_bundledInProgress;

  static void markBundledUpgradeStarted() {
    _bundledInProgress = true;
  }

  static void markBundledUpgradeEnded() {
    _bundledInProgress = false;
  }

  static void setOtaUpgradeInProgress(bool inProgress) {
    _otaInProgress = inProgress;
  }

  /// Test / recovery only.
  static void resetForTest() {
    _bundledInProgress = false;
    _otaInProgress = false;
  }
}
